servicetalk-http-api/src/main/java/io/servicetalk/http/api/RedirectConfigBuilder.java

```diff
@@ -0,0 +1,405 @@
+                                          final CharSequence[] trailersToRedirect,
+                                          final RedirectRequestTransformer userDefinedTransformer) {
+            this.changePostToGet = changePostToGet;
+            this.headersToRedirect = headersToRedirect;
+            this.redirectPayloadBody = redirectPayloadBody;
+            this.trailersToRedirect = trailersToRedirect;
+            this.userDefinedTransformer = userDefinedTransformer;
+        }
+
+        @Override
+        public StreamingHttpRequest apply(final boolean relative,
+                                          final StreamingHttpRequest previousRequest,
+                                          final StreamingHttpResponse redirectResponse,
+                                          final StreamingHttpRequest redirectRequest) {
+            // https://tools.ietf.org/html/rfc7231#section-6.4.2
+            // https://tools.ietf.org/html/rfc7231#section-6.4.3
+            // Note for 301 (Moved Permanently) and 302 (Found):
+            //     For historical reasons, a user agent MAY change the request method from POST to GET for the
+            //     subsequent request.  If this behavior is undesired, the 307 (Temporary Redirect) or
+            //     308 (Permanent Redirect) status codes can be used instead.
+            final int statusCode = redirectResponse.status().code();
+            if (changePostToGet && (statusCode == MOVED_PERMANENTLY.code() || statusCode == FOUND.code()) &&
+                    POST.name().equals(previousRequest.method().name())) {
+                redirectRequest.method(GET);
+            }
+
+            if (relative) {
+                fullCopy(previousRequest, redirectRequest);
+            } else {
+                safeCopy(previousRequest, redirectRequest);
+            }
+            return userDefinedTransformer.apply(relative, previousRequest, redirectResponse, redirectRequest);
+        }
+
+        private static void fullCopy(final StreamingHttpRequest originalRequest,
+                                     final StreamingHttpRequest redirectRequest) {
+            redirectRequest.setHeaders(originalRequest.headers());
+            redirectRequest.transformMessageBody(p -> p.ignoreElements().concat(originalRequest.messageBody()));
+            // Use `transform` to update PayloadInfo flags, assuming trailers may be included in the message body
+            redirectRequest.transform(NOOP_TRAILERS_TRANSFORMER);
+            // FIXME: instead of `transform`, preserve original PayloadInfo/FlushStrategy when it's API is available
+        }
+
+        private void safeCopy(final StreamingHttpRequest request, final StreamingHttpRequest redirectRequest) {
+            // NOTE: for security reasons we do not copy any headers or payload body from the original request by
+            // default for non-relative redirects.
+            for (CharSequence headerName : headersToRedirect) {
```

Comment:
optimization/feel free to ignore for now

if `headersToRedirect.length > request.headers().size()` it may make sense to flip the iteration order, and if that is the case you can also use a `HttpHeaders` as a case-insensitive associating array to do the lookup:

```java
            // cache me in the constructor
            final HttpHeaders hdrsToRedirect = DefaultHttpHeadersFactory.INSTANCE.newHeaders();
            for (CharSequence name : headersToRedirect) {
                hdrsToRedirect.add(name, "anything");
            }
            // cache me in the constructor
            
            for (Map.Entry<CharSequence, CharSequence> hdr : request.headers()) {
                if (hdrsToRedirect.contains(hdr.getKey())) {
                    redirectRequest.addHeader(hdr.getKey(), hdr.getValue());
                }
            }
```